### PR TITLE
fix: pipeline upsert failed to pass parallelism_config to update

### PR DESCRIPTION
**Pull request by qidewenwhen**
*Issue #, if available:* https://github.com/aws/sagemaker-python-sdk/issues/4017

*Description of changes:* 
- pipeline.upsert failed to pass parallelism_config to update, so the updated parallelism_config won't be able to take effect.
- Add more unit/integ tests to prove the functionality of parallelism_config

*Testing done:* unit and integ tests

## Merge Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your pull request._

#### General

- [x] I have read the [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md) doc
- [x] I certify that the changes I am introducing will be backward compatible, and I have discussed concerns about this, if any, with the Python SDK team
- [x] I used the commit message format described in [CONTRIBUTING](https://github.com/aws/sagemaker-python-sdk/blob/master/CONTRIBUTING.md#committing-your-change)
- [x] I have passed the region in to all S3 and STS clients that I've initialized as part of this change.
- [x] I have updated any necessary documentation, including [READMEs](https://github.com/aws/sagemaker-python-sdk/blob/master/README.rst) and [API docs](https://github.com/aws/sagemaker-python-sdk/tree/master/doc) (if appropriate)

#### Tests

- [x] I have added tests that prove my fix is effective or that my feature works (if appropriate)
- [x] I have added unit and/or integration tests as appropriate to ensure backward compatibility of the changes
- [x] I have checked that my tests are not configured for a specific region or account (if appropriate)
- [x] I have used [`unique_name_from_base`](https://github.com/aws/sagemaker-python-sdk/blob/master/src/sagemaker/utils.py#L77) to create resource names in integ tests (if appropriate)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: tests/integ/sagemaker/workflow/test_workflow.py [hunk 1]
```diff
@@ -1019,6 +1020,99 @@ def test_create_and_update_with_parallelism_config(
...
+    )
+
+    try:
+        pipeline.create(role, parallelism_config=dict(MaxParallelExecutionSteps=1))
+        # No ParallelismConfiguration given in pipeline.start, so it won't override that in pipeline.create
+        execution = pipeline.start()
+
+        def validate():
+            # Only one step would be scheduled initially
+            assert len(execution.list_steps()) == 1
```
Comment (svia3) [inline tests/integ/sagemaker/workflow/test_workflow.py:1056]: Just clarifying here -- do we need to check for the `Executing` status here as well? I may be missing what exactly list_steps() is doing. Just taking a brief look at our backend logic, Im seeing that we simply return a page of max 100 results [[link](https://code.amazon.com/packages/IronmanTiogaServiceObjects/blobs/b1b68ec97fe2a8dd94a8be0dd7ae7d459d25ec51/--/src/com/amazon/ironmantiogaservice/datastore/requests/ListPipelineExecutionStepsRequest.java#L20)]

Why would this not be 2, with one in executing status and the other in, I'd assume, `PENDING` or `AVAILABLE_TO_RUN` status?

File: tests/unit/sagemaker/workflow/test_pipeline.py [hunk 2]
```diff
@@ -200,10 +238,12 @@ def test_pipeline_update_with_parallelism_config(sagemaker_session_mock, role_ar
         name="MyPipeline",
         parameters=[],
         steps=[],
-        pipeline_experiment_config=ParallelismConfiguration(max_parallel_execution_steps=10),
```
Comment (svia3) [inline tests/unit/sagemaker/workflow/test_pipeline.py:203]: Was this just incorrect?

Comment (qidewenwhen) [reply tests/unit/sagemaker/workflow/test_pipeline.py:203]: You mean the usage of ParallelismConfiguration? Yea, that won't pass the pipeline definition generation for some reason. Will update

File: tests/integ/sagemaker/workflow/test_workflow.py (hunk 1 above)
Comment (svia3) [reply tests/integ/sagemaker/workflow/test_workflow.py:1056]: Ok I see. So our list response will iteratively grow in size as the execution runs, such that dependent steps will get created as StepEntities in our DDB, and thus show up in the `list_steps()` response page.  Thanks!

File: tests/unit/sagemaker/workflow/test_pipeline.py (hunk 2 above)
Comment (svia3) [reply tests/unit/sagemaker/workflow/test_pipeline.py:203]: Yea, passing that to the `pipeline_experiment_config` in the constructor. Seemed like the incorrect usage originally -- it needs to be passed to `create(parallelism_config=)` call like you already added. Was just checking.

Comment (qidewenwhen) [reply tests/unit/sagemaker/workflow/test_pipeline.py:203]: Ohh, get your point. Yea, the test for parallelism_config was not well-maintained.
Besides of this, the usage of `ParallelismConfiguration` does not work during pipeline definition generation. At this moment, customers and public docs are using dict define the config here. So I just updated accordingly.

**Review by svia3 — APPROVED** [review]

**Review by knikure — APPROVED** [review]

**Pull request merged**